hpx/util/detail/pack_traversal_async_impl.hpp

```diff
@@ -0,0 +1,574 @@
+//  Copyright (c) 2017 Denis Blank
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_UTIL_DETAIL_PACK_TRAVERSAL_ASYNC_IMPL_HPP
+#define HPX_UTIL_DETAIL_PACK_TRAVERSAL_ASYNC_IMPL_HPP
+
+#include <hpx/config.hpp>
+#include <hpx/util/always_void.hpp>
+#include <hpx/util/assert.hpp>
+#include <hpx/util/detail/container_category.hpp>
+#include <hpx/util/detail/pack.hpp>
+#include <hpx/util/invoke.hpp>
+#include <hpx/util/invoke_fused.hpp>
+#include <hpx/util/tuple.hpp>
+
+#include <boost/intrusive_ptr.hpp>
+
+#include <cstddef>
+#include <exception>
+#include <iterator>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace hpx {
+namespace util {
+    namespace detail {
+        /// A tag which is passed to the `operator()` of the visitor
+        /// if an element is visited synchronously.
+        struct async_traverse_visit_tag
+        {
+        };
+        /// A tag which is passed to the `operator()` of the visitor
+        /// if an element is visited after the traversal was detached.
+        struct async_traverse_detach_tag
+        {
+        };
+        /// A tag which is passed to the `operator()` of the visitor
+        /// if the asynchronous pack traversal was finished.
+        struct async_traverse_complete_tag
+        {
+        };
+
+        /// A tag to identify that a mapper shall be constructed in-place
+        /// from the first argument passed.
+        template <typename T>
+        struct async_traverse_in_place_tag
+        {
+        };
+
+        /// Relocates the given pack with the given offset
+        template <std::size_t Offset, typename Pack>
+        struct relocate_index_pack;
+        template <std::size_t Offset, std::size_t... Sequence>
+        struct relocate_index_pack<Offset, pack_c<std::size_t, Sequence...>>
+          : std::common_type<pack_c<std::size_t, (Sequence + Offset)...>>
```

Comment:
Why do we need `common_type` here? The pack is expanded inside of `pack_c`, this looks redundant.

Reply:
Usually I'm using `common_type` as a shorthand alias to provide a `type` alias from the object.
Since `common_type` with a single type is redundant to:

````cpp
template<typename T>
struct type {
  using type = T;
};

Reply:
Make that

Reply:
Not that it matters in this particular instance, but that should be:

    template<typename T>
    struct type {
      using type = std::decay_t<T>;
    };

There doesn't seem to be anything traversal specific to this. Should this functionality be in `util/detail/pack.hpp` instead?

Reply:
@K-ballo yes, maybe we could also move this to `util/detail/pack.hpp` since its not related to the traversal itself.